### DISCUSSION TITLE
it confuses me that how we can extract the non-rectangular images

### DISCUSSION THREAD

**Discussion by BriskyGates** [pymupdf/PyMuPDF#2339] [Q&A] [answered]
```
def extract_images(pdf_path, output_folder):
    doc = fitz.open(pdf_path)
    for page_num in range(len(doc)):
        page = doc.load_page(page_num)
        image_list = page.get_images(full=True)
        for img_index, img in enumerate(image_list):
            xref = img[0]
            base_image = doc.extract_image(xref)
            image_bytes = base_image["image"]

            # save to the png file
            with open(f"{output_folder}/image_p{page_num + 1}_i{img_index + 1}.png", "wb") as img_out:
                img_out.write(image_bytes)
```

[pdf_font_garbled.pdf](https://github.com/pymupdf/PyMuPDF/files/11220342/pdf_font_garbled.pdf)
eg. page2
there exists two kinds of images, the portrait and the watermark, but the former is non-rectangular, how we fill it with black background 

![image_p1_i12](https://user-images.githubusercontent.com/34669414/231714229-7becc278-a0c1-4bc5-854c-499d6655b1de.png)
thanks in advance :>

**Comment by JorjMcKie** [comment, marked as answer]
Looking closer at the images on page 2, you will see that a number of them has masks, these items have a second entry > 0, e.g. `(53, 90, 173, 173, 8, 'DeviceRGB', '', 'Image53', 'DCTDecode')` has 90 there. This is an image mask which must be applied to get the full image. The following snippet only extracts images with a mask and recovers the full picture by applying the mask to the base image:
```python
for item in page.get_images():
    xref = item[0]  # base image xref
    mask = item[1]  # mask xref
    if mask == 0: continue  # ignore if no masked image
    pix0 = fitz.Pixmap(doc, xref)  # pixmap of base image
    if pix0.alpha: pix0 = fitz.Pixmap(pix0, 0)  # remove alpha channel if present
    pixm = fitz.Pixmap(doc, mask)  # pixmap of mask
    pix = fitz.Pixmap(pix0,pixm)  # merge base and mask pixmap to one transparent pixmap
    fname=f"{xref}.png"  # output filename
    pix.save(fname)  # save recovered picture
```
delivers this for base image xref 53:
![53](https://user-images.githubusercontent.com/8290722/231725392-2dfb0c5a-6261-4c3c-b997-3d0d098048c7.png)

**Comment by BriskyGates** [comment]
thanks a lot!!! :>
